spec/ics-25-handler-interface/README.md

```diff
@@ -0,0 +1,314 @@
+---
+ics: 25
+title: Handler Interface
+stage: draft
+category: ibc-core
+author: Christopher Goes <cwgoes@tendermint.com>
+created: 2019-04-23
+modified: 2019-05-03
+---
+
+# Synopsis
+
+This document describes the interface exposed by the standard IBC implementation (referred to as the IBC handler) to modules within the same state machine, and the implementation of that interface by the IBC handler.
+
+# Specification
+
+## Motivation
+
+IBC is an inter-module communication protocol, designed to faciliate reliable, authentication message passing between modules on separate blockchains. Modules should be able to reason about the interface they interact with and the requirements they must adhere to in order to utilize it safely.
+
+## Definitions
+
+`Client` and `RootOfTrust` are as defined in ICS 2.
+
+`Connection` and `ConnectionState` are as defined in ICS 3.
+
+`Channel` is as defined in ICS 4.
+
+`Packet` is as defined in ICS 5.
+
+## Desired Properties
+
+- Client, connection, channel creation as permissionless as possible
+- Dynamic module set
+- Modules can write their own more complex abstractions on top of IBC
+
+## Technical Specification
+
+### Clients
+
+By default, clients are unowned: any module can create a new client, query any existing client, and update any existing client.
+
+`ClientKind` enumerates the list of light client algorithms supported by the handler implementation.
+
+```golang
+type ClientKind enum {
+  Tendermint
+}
+```
+
+`ClientOptions` contains all the parameter choices required to create a client.
+
+```golang
+type ClientOptions struct {
+  ClientKind  kind
+  RootOfTrust rootOfTrust
+}
+```
+
+`ClientInfo` contains information about an existing client.
+
+```golang
+type ClientInfo struct {
+  ClientKind  kind
+  RootOfTrust rootOfTrust
+}
+```
+
+`createClient` creates a new client and returns an automatically allocated identifier.
+
+```coffeescript
+function createClient(ClientOptions options) -> string
+```
+
+`queryClient` queries a client by a known identifier, returning the associated metadata and root of trust if found.
+
+```coffeescript
+function queryClient(string id) -> Maybe<ClientInfo>
+```
+
+`updateClient` updates an existing client with a new header, returning an error if the client was not found or the header was not a valid update.
+
+```coffeescript
+function updateClient(string id, Header header) -> Maybe<Err>
+```
+
+Implementations of `createClient`, `queryClient`, and `updateClient` are defined in ICS 2.
+
+
+### Connections
+
+By default, connections are unowned, but closure is permissioned to the module which created the connection.
+
+`ConnectionKind` enumerates the connection types supported by the handler implementation.
+
+```golang
+type ConnectionKind enum {
+  Transit
```

Comment:
What is "Transit"..."Transmit" maybe? Are "Transit" and "Receive" unidirectional connections?

Reply:
Typo - "Transmit". Yes, they are - those haven't been specified yet, since bidirectional connections are more pressing, but they should be possible.

Fixed the typo.